File: Projects/ReactNativePerfLab/src/models/ProductData.ts
/**
 * Product DTO & UI Wrapper
 * Lightweight TypeScript interfaces for zero-cost JSON deserialization 
 * and deterministic Virtual DOM diffing.
 */

/**
 * The raw data schema representing a product entity.
 * Decoupled from UI-specific identifiers to prevent structural bloat during parsing.
 */
export interface ProductData {
    readonly name: string;
    readonly price: string;
    readonly oldPrice: string;
    readonly discount: string;
    readonly category: string;
    readonly imageName: string;
}

/**
 * A UI-bound wrapper encapsulating the raw product data.
 * Uses a deterministic integer ID to ensure efficient O(1) row reconciliation 
 * in list components without the overhead of string UUIDs.
 */
export interface Product {
    readonly id: number;
    readonly data: ProductData;
}

File: Projects/KotlinMultiPlatformPerfLab/composeApp/src/commonMain/kotlin/com/tez/perflab/models/ProductData.kt
package com.tez.perflab.models

import kotlinx.serialization.Serializable

/**
 * A lightweight Data Transfer Object (DTO) representing a product entity.
 * Optimized for reflection-free JSON deserialization throughput.
 */
@Serializable
data class ProductData(
    val name: String,
    val price: String,
    val oldPrice: String,
    val discount: String,
    val category: String,
    val imageName: String
)

/**
 * A UI-bound wrapper encapsulating the raw product data for rendering pipelines.
 *
 * Uses a deterministic integer ID instead of a dynamic UUID. Generating UUIDs
 * in a tight loop during high-frequency scrolling benchmarks introduces CPU overhead
 * and frequent Garbage Collection (GC). Using a sequential primitive ID isolates
 * pure UI rendering performance.
 */
data class Product(
    val id: Int,
    val data: ProductData
)